File: api/analyze.py
from fastapi import FastAPI, HTTPException, BackgroundTasks
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import FileResponse
from pydantic import BaseModel, HttpUrl
import asyncio
import os
from typing import Optional
import uuid
from datetime import datetime

from analyzer import WebsiteAnalyzer
from report_generator import PDFReportGenerator

app = FastAPI(title="Website Analyzer API", version="1.0.0")

# CORS middleware for frontend communication
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],  # In production, specify your frontend domain
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# In-memory storage for analysis results (in production, use a database)
analysis_results = {}

class AnalysisRequest(BaseModel):
    url: HttpUrl
    options: Optional[dict] = {}

class AnalysisResponse(BaseModel):
    analysis_id: str
    status: str
    message: str

@app.get("/")
async def root():
    return {"message": "Website Analyzer API is running!"}

@app.post("/analyze", response_model=AnalysisResponse)
async def start_analysis(request: AnalysisRequest, background_tasks: BackgroundTasks):
    """Start website analysis in background"""
    analysis_id = str(uuid.uuid4())
    
    # Initialize analysis result
    analysis_results[analysis_id] = {
        "id": analysis_id,
        "url": str(request.url),
        "status": "started",
        "started_at": datetime.now().isoformat(),
        "progress": 0,
        "results": None,
        "error": None
    }
    
    # Start background analysis
    background_tasks.add_task(run_analysis, analysis_id, str(request.url), request.options)
    
    return AnalysisResponse(
        analysis_id=analysis_id,
        status="started",
        message="Analysis started successfully"
    )

@app.get("/analysis/{analysis_id}")
async def get_analysis_status(analysis_id: str):
    """Get analysis status and results"""
    if analysis_id not in analysis_results:
        raise HTTPException(status_code=404, detail="Analysis not found")
    
    return analysis_results[analysis_id]

@app.get("/download/{analysis_id}")
async def download_report(analysis_id: str):
    """Download PDF report"""
    if analysis_id not in analysis_results:
        raise HTTPException(status_code=404, detail="Analysis not found")
    
    result = analysis_results[analysis_id]
    if result["status"] != "completed":
        raise HTTPException(status_code=400, detail="Analysis not completed yet")
    
    pdf_path = f"reports/{analysis_id}.pdf"
    if not os.path.exists(pdf_path):
        raise HTTPException(status_code=404, detail="Report file not found")
    
    return FileResponse(
        path=pdf_path,
        filename=f"website_analysis_{analysis_id[:8]}.pdf",
        media_type="application/pdf"
    )

async def run_analysis(analysis_id: str, url: str, options: dict):
    """Run the complete website analysis"""
    try:
        # Update status
        analysis_results[analysis_id]["status"] = "analyzing"
        analysis_results[analysis_id]["progress"] = 10
        
        # Initialize analyzer
        analyzer = WebsiteAnalyzer()
        
        # Run analysis steps
        analysis_results[analysis_id]["progress"] = 20
        performance_data = await analyzer.analyze_performance(url)
        
        analysis_results[analysis_id]["progress"] = 40
        accessibility_data = await analyzer.analyze_accessibility(url)
        
        analysis_results[analysis_id]["progress"] = 60
        seo_data = await analyzer.analyze_seo(url)
        
        analysis_results[analysis_id]["progress"] = 80
        security_data = await analyzer.analyze_security(url)
        
        analysis_results[analysis_id]["progress"] = 90
        content_data = await analyzer.analyze_content(url)
        
        # Compile results
        results = {
            "url": url,
            "analyzed_at": datetime.now().isoformat(),
            "performance": performance_data,
            "accessibility": accessibility_data,
            "seo": seo_data,
            "security": security_data,
            "content": content_data,
            "overall_score": calculate_overall_score(performance_data, accessibility_data, seo_data, security_data, content_data)
        }
        
        analysis_results[analysis_id]["results"] = results
        analysis_results[analysis_id]["progress"] = 95
        
        # Generate PDF report
        report_generator = PDFReportGenerator()
        pdf_path = await report_generator.generate_report(results, analysis_id)
        
        analysis_results[analysis_id]["status"] = "completed"
        analysis_results[analysis_id]["progress"] = 100
        analysis_results[analysis_id]["pdf_path"] = pdf_path
        
    except Exception as e:
        analysis_results[analysis_id]["status"] = "failed"
        analysis_results[analysis_id]["error"] = str(e)
        print(f"Analysis failed for {analysis_id}: {e}")

def calculate_overall_score(performance, accessibility, seo, security, content):
    """Calculate overall website score"""
    scores = []
    if performance and "score" in performance:
        scores.append(performance["score"])
    if accessibility and "score" in accessibility:
        scores.append(accessibility["score"])
    if seo and "score" in seo:
        scores.append(seo["score"])
    if security and "score" in security:
        scores.append(security["score"])
    if content and "score" in content:
        scores.append(content["score"])
    
    return round(sum(scores) / len(scores)) if scores else 0

# Vercel serverless function handler
def handler(request):
    return app(request.scope, request.receive, request.send)

File: pages/api/analyze.js
import WebsiteAnalyzer from '../../lib/analyzer.js';
import AnalysisStorage from '../../lib/storage.js';
import PDFReportGenerator from '../../lib/reportGenerator.js';
import UserTracker from '../../lib/userTracker.js';

export default async function handler(req, res) {
  // Set CORS headers
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  // Handle CORS preflight
  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  // Only allow POST requests
  if (req.method !== 'POST') {
    res.status(405).json({ error: 'Method not allowed' });
    return;
  }

  try {
    const { url, options = {} } = req.body;

    if (!url) {
      res.status(400).json({ error: 'URL is required' });
      return;
    }

    // Generate analysis ID
    const analysisId = crypto.randomUUID();
    const storage = new AnalysisStorage();
    const analyzer = new WebsiteAnalyzer();
    const userTracker = new UserTracker();

    // Track user and get user info (optional)
    let user = null;
    try {
      user = await userTracker.trackUser(req, {
        name: req.body.name || null,
        email: req.body.email || null,
        contact: req.body.contact || null
      });
    } catch (error) {
      console.log('User tracking failed (optional):', error.message);
    }

    // Initialize analysis result
    const initialResult = {
      id: analysisId,
      url: url,
      status: 'started',
      started_at: new Date().toISOString(),
      progress: 0,
      results: null,
      error: null,
      user_id: user?.id || null
    };

    // Save initial result to local storage
    storage.saveAnalysisResult(analysisId, initialResult);
    
    // Track analysis in Supabase (optional)
    try {
      await userTracker.trackAnalysis(analysisId, url, user?.id, req);
    } catch (error) {
      console.log('Analysis tracking failed (optional):', error.message);
    }

    // Start analysis in background (simplified for serverless)
    try {
      // Update status to analyzing
      initialResult.status = 'analyzing';
      initialResult.progress = 10;
      storage.saveAnalysisResult(analysisId, initialResult);

      // Run analysis steps
      initialResult.progress = 20;
      storage.saveAnalysisResult(analysisId, initialResult);
      const performanceData = await analyzer.analyzePerformance(url);

      initialResult.progress = 40;
      storage.saveAnalysisResult(analysisId, initialResult);
      const accessibilityData = await analyzer.analyzeAccessibility(url);

      initialResult.progress = 60;
      storage.saveAnalysisResult(analysisId, initialResult);
      const seoData = await analyzer.analyzeSEO(url);

      initialResult.progress = 80;
      storage.saveAnalysisResult(analysisId, initialResult);
      const securityData = await analyzer.analyzeSecurity(url);

      initialResult.progress = 90;
      storage.saveAnalysisResult(analysisId, initialResult);
      const contentData = await analyzer.analyzeContent(url);

      // Compile results
      const results = {
        url: url,
        analyzed_at: new Date().toISOString(),
        performance: performanceData,
        accessibility: accessibilityData,
        seo: seoData,
        security: securityData,
        content: contentData,
        overall_score: analyzer.calculateOverallScore(performanceData, accessibilityData, seoData, securityData, contentData)
      };

      initialResult.results = results;
      initialResult.progress = 95;
      storage.saveAnalysisResult(analysisId, initialResult);

      // Generate PDF report
      const reportGenerator = new PDFReportGenerator();
      const pdfBase64 = reportGenerator.generateReport(results, analysisId);

      initialResult.status = 'completed';
      initialResult.progress = 100;
      initialResult.pdf_base64 = pdfBase64;
      storage.saveAnalysisResult(analysisId, initialResult);

      // Update analysis in Supabase with results (optional)
      try {
        await userTracker.updateAnalysis(analysisId, results, pdfBase64);
      } catch (error) {
        console.log('Analysis update failed (optional):', error.message);
      }

    } catch (analysisError) {
      // Update result with error
      initialResult.status = 'failed';
      initialResult.error = analysisError.message;
      storage.saveAnalysisResult(analysisId, initialResult);
    }

    res.status(200).json({
      analysis_id: analysisId,
      status: 'started',
      message: 'Analysis started successfully',
      url: url
    });

  } catch (error) {
    console.error('Error in analyze endpoint:', error);
    res.status(500).json({ error: error.message });
  }
}
